### DISCUSSION TITLE
Special characters missing from font

### DISCUSSION THREAD

**Discussion by meowfishorg** [jgm/pandoc#9334] [Q&A] [answered]
I was running a conversion from markdown to PDF:
```bash
$ pandoc --pdf-engine=xelatex special_chars.md -o output.pdf
[WARNING] Missing character: There is no ❶ (U+2776) (U+2776) in font [lmroman10-regular]:mapping=t
[WARNING] Missing character: There is no 🅐 (U+1F150) (U+1F150) in font [lmroman10-regular]:mapping=
```

The font is  missing, so I tired to use another font:
```bash
pandoc --pdf-engine=xelatex --variable mainfont="Noto Sans Symbols" special_chars.md -o output.pdf
```
Ok, ❶ and 🅐 are now rendered. But some other fonts are not missing...

I also tried another way with HTML in the middle:
```bash
pandoc -t html --pdf-engine=wkhtmltopdf special_chars.md -o output.pdf
```
Ok, the characters are there, but, the PDF is darkened.

Is there a way that I can use a secondary font for missing characters? Or is there supposedly an easy way to resolve this, such as just install X font? I might be missing something. The requirement is to be able to print all fonts correctly, so using other characters is not an option.

**Comment by jgm** [comment]
Easiest approach is to use a font that has all the glyphs you need.

Failing that, you can try to set up a fallback font.  Currently that's a bit complex, though there's a PR to make it easier:
https://github.com/jgm/pandoc/pull/9204
(It only works with lualatex, unfortunately.)

You might also try `-t typst -o output.pdf`. The default font used by typst seems to have just about everything.

**Comment by badumont** [comment]
If you stay with XeLaTeX, you can setup the font to use for every missing character, as BPJ explained on the (now inaccessible) mailing list:

  The easiest fix is to put this in a text file headerincludes.tex:

  \usepackage{newunicodechar}

  \newfontface\notoemoji{Noto Emoji}

  \newunicodechar{📁}{\notoemoji 📁} 

  Repeat the last line substituting each emoji you want to use.

  You may need to use Noto Sans Symbols for any non-emoji symbols you
  want to use which XeLaTeX complains about. Just repeat the second line
  suitably adjusted, and then use similar lines to the third line with
  the font command in the second pair of braces adjusted.

  Then when you run pandoc add "-H headerincludes.tex" (without the
  quotes) to the command line.

**Reply by meowfishorg** [reply]
I did this:
```tex
\usepackage{newunicodechar}

\newfontface\notosanssymbols{Noto Sans Symbols}

\newunicodechar{❶}{\notosanssymbols ❶}
\newunicodechar{❷}{\notosanssymbols ❷}
\newunicodechar{❸}{\notosanssymbols ❸}
\newunicodechar{❹}{\notosanssymbols ❹}
\newunicodechar{❺}{\notosanssymbols ❺}
\newunicodechar{❻}{\notosanssymbols ❻}
\newunicodechar{❼}{\notosanssymbols ❼}
\newunicodechar{❽}{\notosanssymbols ❽}
\newunicodechar{❾}{\notosanssymbols ❾}
```

Your method works, but the usual punctuation characters are now missing:
```txt
[WARNING] Missing character: There is no . (U+002E) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no . (U+002E) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no , (U+002C) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no . (U+002E) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no . (U+002E) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no ` (U+0060) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no ` (U+0060) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no ' (U+0027) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no ' (U+0027) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no . (U+002E) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no , (U+002C) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no , (U+002C) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no . (U+002E) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no , (U+002C) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no , (U+002C) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no . (U+002E) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no ` (U+0060) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no ` (U+0060) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no ' (U+0027) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no ' (U+0027) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no ` (U+0060) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no ` (U+0060) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no ' (U+0027) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no ' (U+0027) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no ( (U+0028) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no ) (U+0029) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no . (U+002E) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no ( (U+0028) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no ) (U+0029) in font Noto Sans Symbols Regular/OT:
[WARNING] Missing character: There is no . (U+002E) in font Noto Sans Symbols Regular/OT:
```

I'm quite new to this, so I must be missing something. Any guidance? Thank you.

**Comment by badumont** [comment, marked as answer]
I think that e.g.  "❶ first list item." expands to "\notosanssymbols ❶ first list item.", so that \notosanssymbols remains in effect after the ❶ character. Doubling the braces in the definitions (e.g. "\newunicodechar{❶}{{\notosanssymbols ❶}}") should fix the problem.

**Reply by meowfishorg** [reply]
Works perfectly! Thank you.
